### PR TITLE
Fetch data home screen

### PR DESCRIPTION
**Pull request by nhatdaoquang25**

### CODE REVIEW

File: lib/src/services/coin_service/app_coin_service.dart [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+  @override
+  Future<List<Coins>> fecthCoins() async {
+    Map<String, String> qParams = {
+      AppConfig.instance.getValue(AppConstants.currency): 'usd',
+      AppConfig.instance.getValue(AppConstants.order): 'market_cap_desc',
+      AppConfig.instance.getValue(AppConstants.perPage): '20',
+      AppConfig.instance.getValue(AppConstants.page): '1',
+      AppConfig.instance.getValue(AppConstants.sparkline): 'false',
+    };
```
Comment (HaiNguyenPV12) [inline lib/src/services/coin_service/app_coin_service.dart]: Should move default parameters into constants or config file

File: lib/src/widgets/custom_card.dart [hunk 2]
```diff
@@ -46,11 +46,14 @@ class CustomCard extends StatelessWidget {
...
-                  Padding(
-                    padding: const EdgeInsets.all(12.0),
-                    child: Image.network(
-                      imageNetwork,
-                      fit: BoxFit.contain,
+                  SizedBox(
+                    width: 50,
+                    height: 50,
+                    child: Padding(
+                      padding: const EdgeInsets.all(12.0),
```
Comment (HaiNguyenPV12) [inline lib/src/widgets/custom_card.dart]: should turn these values into variables

File: lib/src/screens/home_screen/home_screen.dart [hunk 3]
```diff
@@ -35,42 +48,73 @@ class HomeScreen extends StatelessWidget {
...
+          }
+          if (state is CoinLoadFailure) {
+            return Container(
+              color: Colors.red,
+              alignment: Alignment.center,
+              child: Text(state.errorMessage!),
+            );
+          }
+          if (state is CoinLoadSucess) {
+            var coin = state.coins;
```
Comment (HaiNguyenPV12) [inline lib/src/screens/home_screen/home_screen.dart]: why is the variable named `coin` when the state property is `coins`?

**Review by HaiNguyenPV12 — CHANGES_REQUESTED** [review]

**Comment by HaiNguyenPV12** [comment]
Also, unit tests?

File: lib/src/services/coin_service/app_coin_service.dart (hunk 1 above)
Comment (nhatdaoquang25) [reply lib/src/services/coin_service/app_coin_service.dart]: I will change it.

File: lib/src/widgets/custom_card.dart (hunk 2 above)
Comment (nhatdaoquang25) [reply lib/src/widgets/custom_card.dart]: I will change it to variables

File: lib/src/screens/home_screen/home_screen.dart (hunk 3 above)
Comment (nhatdaoquang25) [reply lib/src/screens/home_screen/home_screen.dart]: I will change the named

**Comment by nhatdaoquang25** [comment]
> Also, unit tests?
In my plan, the next week will be test everything. So, I code the project first.

File: lib/src/routes/route_genertor.dart [hunk 4]
```diff
@@ -1,14 +1,27 @@
...
+
+import '/../src/blocs/coin_bloc/coin_bloc.dart';
+import '/../src/services/coin_service/app_coin_service.dart';
+
 import '/../src/screens/detail_screen/detail_screen.dart';
 import '/../src/screens/search_screen/search_screen.dart';
 import '/../src/constants/name_routes_constants.dart';
 import '/../src/screens/home_screen/home_screen.dart';
 
+import 'package:http/http.dart' as http;
```
Comment (thanhphamtrung) [inline lib/src/routes/route_genertor.dart]: sort

File: lib/src/routes/route_genertor.dart [hunk 5]
```diff
@@ -1,14 +1,27 @@
...
+    final httpClient = http.Client();
     switch (settings.name) {
       case NameRoutesConstants.root:
-        return MaterialPageRoute(builder: (_) => const HomeScreen());
+        return MaterialPageRoute(
+            builder: (_) => BlocProvider(
+                  create: (context) =>
+                      CoinBloc(coinService: AppAlbumService(httpClient)),
+                  child: const HomeScreen(),
+                ));
```
Comment (thanhphamtrung) [inline lib/src/routes/route_genertor.dart]: use multi bloc provider in app.dart for easier to manage

File: lib/src/screens/home_screen/home_screen.dart [hunk 6]
```diff
@@ -1,16 +1,29 @@
...
 
+  @override
+  State<HomeScreen> createState() => _HomeScreenState();
+}
+
+class _HomeScreenState extends State<HomeScreen> {
   @override
   Widget build(BuildContext context) {
+    final percentageFormat = intl.NumberFormat("##0.0#");
+    context.read<CoinBloc>().add(CoinRequested());
```
Comment (thanhphamtrung) [inline lib/src/screens/home_screen/home_screen.dart]: move this line to initState if you do not need to call it multiple times

File: lib/src/screens/home_screen/home_screen.dart [hunk 7]
```diff
@@ -35,42 +48,71 @@ class HomeScreen extends StatelessWidget {
...
+                  child: Column(
+                    children: [
+                      Expanded(
+                        flex: 1,
+                        child: Container(),
+                      ),
+                      Expanded(
+                        flex: 9,
+                        child: ListView.builder(
+                          padding: const EdgeInsets.only(top: 20),
```
Comment (thanhphamtrung) [inline lib/src/screens/home_screen/home_screen.dart]: layout variable

File: lib/src/screens/home_screen/home_screen.dart [hunk 8]
```diff
@@ -35,42 +48,71 @@ class HomeScreen extends StatelessWidget {
...
+                          itemCount: state.coins!.length,
+                          itemBuilder: (context, index) {
+                            var coinIndex = state.coins![index];
+                            return Padding(
+                                padding: const EdgeInsets.only(
+                                    left: 10, right: 10, bottom: 5),
```
Comment (thanhphamtrung) [inline lib/src/screens/home_screen/home_screen.dart]: layout variable

File: lib/src/screens/home_screen/home_screen.dart [hunk 9]
```diff
@@ -35,42 +48,71 @@ class HomeScreen extends StatelessWidget {
...
+                        flex: 1,
+                        child: Container(),
+                      ),
+                      Expanded(
+                        flex: 9,
+                        child: ListView.builder(
+                          padding: const EdgeInsets.only(top: 20),
+                          itemCount: state.coins!.length,
+                          itemBuilder: (context, index) {
+                            var coinIndex = state.coins![index];
```
Comment (thanhphamtrung) [inline lib/src/screens/home_screen/home_screen.dart]: what happens if state.coins is null

File: lib/src/screens/home_screen/home_screen.dart [hunk 10]
```diff
@@ -35,42 +48,71 @@ class HomeScreen extends StatelessWidget {
...
+                                  index: index,
+                                  onTap: () {
+                                    Navigator.of(context).pushNamed(
+                                        NameRoutesConstants.detailScreen);
+                                  },
+                                  name: coinIndex.name,
+                                  imageNetwork: coinIndex.image,
+                                  currentPrice: coinIndex.currentPrice,
+                                  priceChange24h: num.parse(percentageFormat
+                                      .format(coinIndex.priceChange24H)),
```
Comment (thanhphamtrung) [inline lib/src/screens/home_screen/home_screen.dart]: what happens if num.parse return exception

File: lib/src/services/coin_service/app_coin_service.dart [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+import '/../src/models/coins.dart';
+import '/../src/services/coin_service/coin_service.dart';
+
+class AppAlbumService extends CoinService {
+  AppAlbumService(http.Client client) : super(client);
+
+  @override
+  Future<List<Coins>> fecthCoins() async {
+    Map<String, String> qParams = {
```
Comment (thanhphamtrung) [inline lib/src/services/coin_service/app_coin_service.dart]: qParams what does it mean?

File: lib/src/services/coin_service/app_coin_service.dart [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+            AppConfig.instance.getValue(AppConstants.marketPath),
+        queryParameters: qParams);
+
+    var response = await client.get(uri);
+
+    if (response.statusCode == 200) {
+      Iterable responseList = json.decode(response.body);
+
+      var coins =
+          List<Coins>.from(responseList.map((model) => Coins.fromJson(model)));
```
Comment (thanhphamtrung) [inline lib/src/services/coin_service/app_coin_service.dart]: what happens if fromJson return exception

File: lib/src/widgets/custom_card.dart [hunk 13]
```diff
@@ -61,28 +66,44 @@ class CustomCard extends StatelessWidget {
...
+                                      Icons.arrow_drop_down_sharp,
+                                      color: Colors.red,
+                                    ),
+                                    Text(
+                                      '$priceChange24h ($priceChangePercentage24h%)',
+                                      style: const TextStyle(
+                                          fontSize: 16, color: Colors.red),
+                                    ),
+                                  ],
+                                )
```
Comment (thanhphamtrung) [inline lib/src/widgets/custom_card.dart]: what differences between these Row?

File: lib/src/services/coin_service/app_coin_service.dart (hunk 11 above)
Comment (nhatdaoquang25) [reply lib/src/services/coin_service/app_coin_service.dart]: That's queryParameters. I will change the name.

File: lib/src/screens/home_screen/home_screen.dart (hunk 8 above)
Comment (nhatdaoquang25) [reply lib/src/screens/home_screen/home_screen.dart]: Yah, I will bring it to variable

File: lib/src/screens/home_screen/home_screen.dart (hunk 7 above)
Comment (nhatdaoquang25) [reply lib/src/screens/home_screen/home_screen.dart]: Yah, I will bring it to variable

File: lib/src/screens/home_screen/home_screen.dart (hunk 6 above)
Comment (nhatdaoquang25) [reply lib/src/screens/home_screen/home_screen.dart]: I move it to class Route. And after that I add it after CoinBloc.

File: lib/src/routes/route_genertor.dart (hunk 5 above)
Comment (nhatdaoquang25) [reply lib/src/routes/route_genertor.dart]: Yah, I got it

File: lib/src/routes/route_genertor.dart (hunk 4 above)
Comment (nhatdaoquang25) [reply lib/src/routes/route_genertor.dart]: That's mean sort to A-Z name.

File: lib/src/services/coin_service/app_coin_service.dart (hunk 12 above)
Comment (nhatdaoquang25) [reply lib/src/services/coin_service/app_coin_service.dart]: Can I use try catch to it ?

File: lib/src/config/constants.dart [hunk 14]
```diff
@@ -1,7 +1,19 @@
...
+  static const String sparkline = 'SPARKLINE';
+  static const String order = 'ORDER';
+  static const String defaultEnv = 'production';
+  static const String envConfigPath = 'assets/config/env/';
+  static const String envConfigType = '.json';
+  static const String currencyOfMarket = 'usd';
+  static const String orderBy = 'market_cap_desc';
+  static const String amountPerPage = '20';
+  static const String numberPage = '1';
+  static const String valueSparkline = 'false';
```
Comment (HaiNguyenPV12) [inline lib/src/config/constants.dart]: boolean variables should start with `is`

**Review by HaiNguyenPV12 — DISMISSED** [review]

File: lib/src/config/constants.dart (hunk 14 above)
Comment (nhatdaoquang25) [reply lib/src/config/constants.dart]: I will change it

File: coverage/lcov.info [hunk 15]
```diff
@@ -0,0 +1,326 @@
+SF:lib\src\blocs\coin_bloc\coin_bloc.dart
```
Comment (thanhphamtrung) [inline coverage/lcov.info]: do not push this file

File: lib/src/app.dart [hunk 16]
```diff
@@ -1,25 +1,33 @@
...
+import 'package:flutter_bloc/flutter_bloc.dart';
+import '/../src/blocs/coin_bloc/coin_bloc.dart';
+import '/../src/services/coin_service/coin_service_impl.dart';
 
 import '/../src/constants/name_routes_constants.dart';
-import '/../src/constants/string_constants.dart';
 import '/../src/routes/route_genertor.dart';
+import 'package:http/http.dart' as http;
+
+import 'blocs/search_bloc/search_bloc.dart';
```
Comment (thanhphamtrung) [inline lib/src/app.dart]: sort

File: lib/src/blocs/coin_bloc/coin_bloc.dart [hunk 17]
```diff
@@ -1,10 +1,33 @@
 import 'package:flutter_bloc/flutter_bloc.dart';
+import '../../models/coins.dart';
 
 import '/../src/blocs/coin_bloc/coin_event.dart';
 import '/../src/blocs/coin_bloc/coin_state.dart';
+import '/../src/services/coin_service/coin_service.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/coin_bloc/coin_bloc.dart]: sort

File: lib/src/blocs/coin_bloc/coin_event.dart [hunk 18]
```diff
@@ -1 +1,18 @@
...
+  final int numberPage;
+
+  CoinRequested({required this.numberPage});
+}
+
+class CoinLoadMore extends CoinEvent {
+  final int numberPage;
+
+  CoinLoadMore({required this.numberPage});
+}
```
Comment (thanhphamtrung) [inline lib/src/blocs/coin_bloc/coin_event.dart]: why did you this override?
 @override
  List<Object?> get props => [];

File: lib/src/blocs/search_bloc/search_bloc.dart [hunk 19]
```diff
@@ -0,0 +1,23 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+import '../../services/coin_service/coin_service.dart';
+
+import 'search_event.dart';
+import 'search_state.dart';
+
+class SearchBloc extends Bloc<SearchEvent, SearchState> {
+  final CoinService? coinService;
```
Comment (thanhphamtrung) [inline lib/src/blocs/search_bloc/search_bloc.dart]: can it be null?

File: lib/src/blocs/search_bloc/search_state.dart [hunk 20]
```diff
@@ -0,0 +1,27 @@
+import 'package:equatable/equatable.dart';
+import '../../models/coins.dart';
```
Comment (thanhphamtrung) [inline lib/src/blocs/search_bloc/search_state.dart]: line break

File: lib/src/screens/home_screen/home_screen.dart [hunk 21]
```diff
@@ -1,14 +1,50 @@
...
+  State<HomeScreen> createState() => _HomeScreenState();
+}
+
+class _HomeScreenState extends State<HomeScreen> {
+  final percentageFormat = intl.NumberFormat("##0.0#");
+
+  double paddingTopOfListView = 20;
+  double paddingLeftOfCustomCard = 10;
+  double paddingRightOfCustomCard = 10;
+  double paddingBottomtOfCustomCard = 5;
```
Comment (thanhphamtrung) [inline lib/src/screens/home_screen/home_screen.dart]: make them private

File: lib/src/screens/search_screen/search_screen.dart [hunk 22]
```diff
@@ -1,17 +1,39 @@
...
+    double sizeWidth = MediaQuery.of(context).size.width;
+
+    double spacerTop = sizeHeight / 10;
+    double maxHeight = sizeHeight / 2;
+    double sizedBoxWidth = sizeWidth / 12;
+
+    double borderRaidus = 10.0;
+
+    double paddingLeft = 5;
+    double paddingRight = 25;
```
Comment (thanhphamtrung) [inline lib/src/screens/search_screen/search_screen.dart]: think a little for more variable meaning

File: lib/src/widgets/custom_card.dart [hunk 23]
```diff
@@ -46,11 +48,14 @@ class CustomCard extends StatelessWidget {
...
-                  Padding(
-                    padding: const EdgeInsets.all(12.0),
-                    child: Image.network(
-                      imageNetwork,
-                      fit: BoxFit.contain,
+                  SizedBox(
+                    width: widthSizedBox,
+                    height: heightSizedBox,
+                    child: Padding(
+                      padding: const EdgeInsets.all(10.0),
```
Comment (thanhphamtrung) [inline lib/src/widgets/custom_card.dart]: .

**Review by thanhphamtrung — COMMENTED** [review]
layout variable
sort import x2
push coverage file?
=> 40

File: lib/src/screens/home_screen/home_screen.dart (hunk 21 above)
Comment (nhatdaoquang25) [reply lib/src/screens/home_screen/home_screen.dart]: I fixed it

File: lib/src/blocs/search_bloc/search_state.dart (hunk 20 above)
Comment (nhatdaoquang25) [reply lib/src/blocs/search_bloc/search_state.dart]: I fixed it :(

File: lib/src/blocs/coin_bloc/coin_event.dart (hunk 18 above)
Comment (nhatdaoquang25) [reply lib/src/blocs/coin_bloc/coin_event.dart]: I will add it

File: lib/src/blocs/coin_bloc/coin_bloc.dart (hunk 17 above)
Comment (nhatdaoquang25) [reply lib/src/blocs/coin_bloc/coin_bloc.dart]: I fixed it :(

File: lib/src/app.dart (hunk 16 above)
Comment (nhatdaoquang25) [reply lib/src/app.dart]: I fixed it :(

File: coverage/lcov.info (hunk 15 above)
Comment (nhatdaoquang25) [reply coverage/lcov.info]: I will remove that

**Review by thanhphamtrung — DISMISSED** [review]

**Review by thanhphamtrung — APPROVED** [review]
LGTM

**Pull request merged**